docs/design/paginators.md

```diff
@@ -70,127 +69,113 @@ used in a model.
+The function generated always begins with the prefix `paginate`, helping with the `discoverablity` design consideration, by allowing users to quickly 
+view all paginatable operations by specifing this literal prefix in the IDE.  In a similar way, extension functions that provide direct access to a paginator's
+nested member will always begin with the prefix `on`.  Once customers are familiar with these two prefixes, discovering what pagination options are available for a given
+client or paginated response is simple and consistent.
 
-```kotlin
-class ListFunctionsPaginator(
-    private val client: LambdaClient,
-    private val initialRequest: ListFunctionsRequest
-): SdkAsyncIterable<ListFunctionsResponse>, SdkAsyncIterator<ListFunctionsResponse> {
-   private var cursor: String? = null
-   private var isFirstPage: Boolean = true
+Using the model in the introduction would produce the following:
 
-   override operator fun iterator(): SdkAsyncIterator<ListFunctionsResponse> = this
+### Examples
 
-   override suspend operator fun hasNext(): Boolean
-           = isFirstPage || (cursor?.isNotEmpty() ?: false)
+#### Usage - Manual Pagination
 
-   override suspend operator fun next(): ListFunctionsResponse {
-      if (!hasNext()) throw NoSuchElementException("no pages remaining")
+An example of driving a paginator manually and processing results:
 
-      val req = initialRequest.copy {
-         this.marker = cursor
+```kotlin
+suspend fun rawPaginationExample(client: LambdaClient) {
+   lambdaClient
+      .paginateListFunctions(ListFunctionsRequest {})
+      .collect { response ->
+         response.functions?.forEach { functionConfiguration ->
+            println(functionConfiguration.functionName) 
+         }
       }
-
-      val result = client.listFunctions(req)
-      isFirstPage = false
-      cursor = result.nextMarker
-      return result
-   }
-
-   // "functions" is the member targeted by the trait's `items` property
-   val functions: SdkAsyncIterable<FunctionConfiguration> = mapNotNull(ListFunctionsResponse::functions).flatten()
 }
 ```
 
-The shape targeted by `inputToken` is mapped to the `cursor` field. Each iteration the `cursor` is updated with the output field targeted by the `outputToken`.
-
+#### Usage - Iterating over Modeled Item
 
-### Creating a Paginator
-
-Each operation that supports pagination would receive an extension function to create a paginator. Using the model in the introduction would produce the following:
+Notice that this example is a superset of the previous.  Simply by providing a transform we are able to iterate over the 
+nested element.  A key design feature of this approach is that users may extend the abstraction as needed simply by providing
+their own flow transforms.  
 
+Additionally, the `item` specified in the model may be deeply nested and involve operations and types with many words.  Generating a
```

Comment:
I see. Do you have an example of this? It would be helpful to clarify with a real example. I'm also wondering if we just use the leaf member targeted by `items`

Reply:
From grepping the models:

```
./redshift.2012-12-01.json:                    "items": "DefaultClusterParameters.Parameters",
./configservice.2014-11-12.json:                    "items": "OrganizationConformancePackDetailedStatuses",
./servicequotas.2019-06-24.json:                    "items": "ServiceQuotaIncreaseRequestInTemplateList",
./ec2.2016-11-15.json:                    "items": "LocalGatewayRouteTableVirtualInterfaceGroupAssociations",
```
I removed the part about the nesting, from what I see at most there are two levels of testing.  It's just the long winded "composite nouns" that present readability issues in codegen afaik.

Reply:
Indeed those are some long names even without nesting. Either proposal is going to yield long function names. 

Here is an example from Java v2 for the `LocalGatewayRouteTableVirtualInterfaceGroupAssociations`
* [EC2 Paginator Factory Method](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/ec2/Ec2Client.html#describeLocalGatewayVirtualInterfaceGroupsPaginator-software.amazon.awssdk.services.ec2.model.DescribeLocalGatewayVirtualInterfaceGroupsRequest-)
* [EC2 Paginator Type](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/ec2/paginators/DescribeLocalGatewayRouteTableVirtualInterfaceGroupAssociationsIterable.html#localGatewayRouteTableVirtualInterfaceGroupAssociations--)

And here is the one for Config Service `OrganizationConformancePackDetailedStatuses`
* [Config Paginator Type](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/config/paginators/GetOrganizationConformancePackDetailedStatusIterable.html)

Reply:
As mentioned above, design has been updated.  The proposal is now to generate the string literal `items` as the function name to access the iterating value.